### PR TITLE
Support layer LSTM

### PR DESCRIPTION
**Pull request by Vimalakirti**
This PR makes the following changes for proving and verifying an LSTM:
- `onnx.rs`: support multiple outputs from a layer and filter out optional input tags (which are "")
- Add layers and basicblocks for `ConstantOfShape`, `Split`, `LSTM`, `Tanh`, `Sigmoid`

But currently this PR only supports LSTM with `batch_size=1` and `seq_len=1`.

(LSTM layer has been successfully proven and verified. The onnx for testing the LSTM layer is in https://drive.google.com/file/d/13RsN1x5m5JGd0PvBBSSVbOzNF8gzGB2q/view?usp=drive_link)

### CODE REVIEW

File: src/basic_block/split.rs
Comment (liliatangxy) [inline src/basic_block/split.rs]: I think we need to change `!=` to `<`

File: src/basic_block/split.rs
Comment (liliatangxy) [inline src/basic_block/split.rs]: I think this supports `self.axis < inputs[0].ndim() - 1`

File: src/onnx.rs
Comment (liliatangxy) [inline src/onnx.rs]: What node inputs are `""`?

**Comment by Vimalakirti** [comment]
This PR is now ready for reviews. It introduces several new layers for proving LSTM and modifies some old ones:
- Ceil, Sigmoid, Tanh: implemented this by nonlinear macro
- ConstantOfShape: implemented ConstOfShapeBasicBlock
- LSTM: now it supports arbitrary sequence length
- Range: the range check relies on CQ
- ScatterND: by CopyConstraint
- Slice: by CopyConstraint
- Split:  when split along the last axis, we perform Permute first. (like reshape layer)
- Equal: modified old one and performed nonzero check with CQ (modified the setup field of CQ for this purpose)

Besides, some files are modified because of formatting.

File: src/basic_block/cq.rs
Comment (liliatangxy) [inline src/basic_block/cq.rs]: Is this used to support non-contiguous ranges? I think we can make the type be a 1d array or a vec

File: src/basic_block/cq.rs
Comment (liliatangxy) [inline src/basic_block/cq.rs]: remove comment

File: src/layer/lstm.rs
Comment (liliatangxy) [inline src/layer/lstm.rs]: is the comment up to date

File: src/layer/lstm.rs
Comment (liliatangxy) [inline src/layer/lstm.rs:54]: I think for now this is ok but we should raise an issue for proving this case of Split since it is currently annoying to do with permute/copy constraint due to producing multiple outputs

File: src/layer/lstm.rs
Comment (liliatangxy) [inline src/layer/lstm.rs]: is there a reason to have these as two separate lines? readability-wise i think it's ok without since the node var name and the comment make it clear

**Review by liliatangxy — APPROVED** [review]

File: src/layer/lstm.rs
Comment (Vimalakirti) [reply src/layer/lstm.rs:54]: The issue is created here: 
https://github.com/uiuc-kang-lab/zk-torch/issues/54

File: src/onnx.rs
Comment (Vimalakirti) [reply src/onnx.rs]: I remember it's a Slice node in RNNT. An optional field that is supposed to be named "steps" (or something else) turns out to be named "", and results in graph compilation errors. Because "" only happens when the field value is not provided, I think we can directly filter it out to prevent possible errors.

File: src/layer/lstm.rs
Comment (Vimalakirti) [reply src/layer/lstm.rs]: yes, it's up to date. We do not support peepholes and do not support different sequence lengths in a batch. I removed "different sequence lengths" because we require batch_size=1 now.

File: src/basic_block/constant.rs
Comment (ddkang) [inline src/basic_block/constant.rs:45]: I don't know what this is

File: src/basic_block/eq.rs
Comment (ddkang) [inline src/basic_block/eq.rs:52]: Comments

File: src/basic_block/range.rs
Comment (ddkang) [inline src/basic_block/range.rs:12]: Does this need to be proved to be secure?

File: src/layer/arithmetic.rs
Comment (ddkang) [inline src/layer/arithmetic.rs:11]: Why was this file changed?

File: src/layer/constantofshape.rs
Comment (ddkang) [inline src/layer/constantofshape.rs:10]: Don't know what this does

File: src/layer/equal.rs
Comment (ddkang) [inline src/layer/equal.rs:12]: Why did this get so much more complex?

Comment (ddkang) [reply src/layer/equal.rs:12]: It seems we should special case some easy cases for high performance

**Review by ddkang — COMMENTED** [review]
Please wait for my review before merging in the future, especially for large PRs like this

File: src/layer/arithmetic.rs
Comment (Vimalakirti) [reply src/layer/arithmetic.rs:11]: It changed because of formatting. The contents were not changed.

File: src/layer/equal.rs
Comment (Vimalakirti) [reply src/layer/equal.rs:12]: The original EqualLayer returns A==1, and it requires all elements in A must be 1 or 0. The updated EqualLayer returns A==B, which completely implements the ONNX Equal. 
I agree considering simple cases can improve our performance, the issue is created here:
https://github.com/uiuc-kang-lab/zk-torch/issues/56

File: src/basic_block/constant.rs
Comment (Vimalakirti) [reply src/basic_block/constant.rs:45]: Responded here:
https://github.com/uiuc-kang-lab/zk-torch/pull/57#pullrequestreview-2176836918

File: src/basic_block/eq.rs
Comment (Vimalakirti) [reply src/basic_block/eq.rs:52]: Responded here:
https://github.com/uiuc-kang-lab/zk-torch/pull/57#pullrequestreview-2176836918

File: src/basic_block/range.rs
Comment (Vimalakirti) [reply src/basic_block/range.rs:12]: Responded here:
https://github.com/uiuc-kang-lab/zk-torch/pull/57#pullrequestreview-2176836918

File: src/layer/constantofshape.rs
Comment (Vimalakirti) [reply src/layer/constantofshape.rs:10]: Responded here:
https://github.com/uiuc-kang-lab/zk-torch/pull/57#pullrequestreview-2176836918

**Pull request merged**
